File: scripts/generate-svg-sprite-sheet.ts
import { join } from '@std/path/join';
import { basename } from '@std/path/basename';

const INPUT_FOLDER = './src/website/static/svg';
const OUTPUT_FILE = './src/website/static/svg/sprite_sheet.svg';

const SVG_SPRITE_HEADER = `<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">\n`;
const SVG_SPRITE_FOOTER = `\n</svg>`;

async function generateSvgSpriteSheet() {
    try {
        const spriteSymbols: string[] = [];

        for await (const dirEntry of Deno.readDir(INPUT_FOLDER)) {
            if (dirEntry.isFile && dirEntry.name.endsWith('.svg')) {
                const filePath = join(INPUT_FOLDER, dirEntry.name);

                const svgContent = await Deno.readTextFile(filePath);

                const iconId = basename(dirEntry.name, '.svg');

                const pathMatch = svgContent.match(/<path\s+d="([^"]+)"/);

                if (pathMatch) {
                    const pathData = pathMatch[1];

                    const symbol = `  <symbol id="${iconId}" viewBox="0 -960 960 960">
    <path d="${pathData}" />
  </symbol>`;

                    spriteSymbols.push(symbol);
                } else {
                    console.warn(`Could not find a <path> or viewBox in ${dirEntry.name}`);
                }
            }
        }

        spriteSymbols.sort();
        const finalSpriteContent = SVG_SPRITE_HEADER + spriteSymbols.join('\n\n') +
            SVG_SPRITE_FOOTER;

        await Deno.writeTextFile(OUTPUT_FILE, finalSpriteContent);

        console.log(`Successfully created SVG sprite sheet at ${OUTPUT_FILE}`);
        console.log(`Contains ${spriteSymbols.length} icons.`);
    } catch (error) {
        console.error('An error occurred:', error);
    }
}

generateSvgSpriteSheet();

File: src/scripts/generate_svg_sprite_sheet.go
package main

import (
	"fmt"
	"log"
	"os"
	"path"
	"regexp"
	"sort"
	"strings"
)

var inputFolder = path.Join("src", "app", "website", "static", "svg")
var outputFile = path.Join("src", "app", "website", "static", "svg", "sprite_sheet.svg")

var svgSpriteHeader = `<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">` + "\n"
var svgSpriteFooter = "\n" + `</svg>`

func GenerateSvgSpriteSheet() {
	entries, err := os.ReadDir(inputFolder)
	sprites := []string{}
	Fatal(err)
	for _, entry := range entries {
		if strings.Contains(entry.Name(), "sprite_sheet.svg") {
			continue
		}
		filePath := path.Join(inputFolder, entry.Name())
		svgContentBytes, err := os.ReadFile(filePath)
		Fatal(err)
		//svgContent := string(svgContentBytes)
		iconId := strings.TrimSuffix(entry.Name(), ".svg")
		matcher := regexp.MustCompile(`<path\s+d="([^"]+)"`)
		match := matcher.FindSubmatch(svgContentBytes)
		if len(match) > 1 {
			pathData := match[1]
			symbol := fmt.Sprintf(
				`  <symbol id="%s" fill="currentColor" viewBox="0 -960 960 960">
    <path d="%s" />
  </symbol>`, iconId, pathData)
			sprites = append(sprites, symbol)
		} else {
			log.Println("WARN: Could not find a <path> or viewBox in", filePath)
		}
	}

	sort.Strings(sprites)
	finalSpriteContent := svgSpriteHeader + strings.Join(sprites, "\n\n") + svgSpriteFooter

	os.WriteFile(outputFile, []byte(finalSpriteContent), 0o666)
}
